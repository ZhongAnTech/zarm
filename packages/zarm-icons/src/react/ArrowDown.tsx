import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowDown = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M494.094 733.426a41.472 41.472 0 01-20.023-11.126L140.16 388.388c-16.272-16.271-16.272-42.653 0-58.925 16.271-16.272 42.653-16.272 58.925 0L503.66 634.037 808.186 329.51c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L533.2 722.347c-10.607 10.607-25.508 14.3-39.106 11.08z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowDown);
export default ForwardRef;
