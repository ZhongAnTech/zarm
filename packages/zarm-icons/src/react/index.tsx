export { default as ArrowDown } from './ArrowDown';
export { default as ArrowLeft } from './ArrowLeft';
export { default as ArrowRight } from './ArrowRight';
export { default as ArrowUp } from './ArrowUp';
export { default as CloseCircleFill } from './CloseCircleFill';
export { default as CloseCircle } from './CloseCircle';
export { default as Close } from './Close';
export { default as DeleteKey } from './DeleteKey';
export { default as Empty } from './Empty';
export { default as Keyboard } from './Keyboard';
export { default as Minus } from './Minus';
export { default as PlusCircle } from './PlusCircle';
export { default as Plus } from './Plus';
export { default as Search } from './Search';
export { default as SuccessCircle } from './SuccessCircle';
export { default as Success } from './Success';
export { default as Volume } from './Volume';
export { default as WarningCircle } from './WarningCircle';
export { default as Warning } from './Warning';
