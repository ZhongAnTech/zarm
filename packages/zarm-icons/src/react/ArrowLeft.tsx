import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowLeft = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M296.114 508.035c-3.22-13.597.473-28.499 11.079-39.105l333.912-333.912c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L395.504 498.47l304.574 304.574c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L307.241 528.058a41.472 41.472 0 01-11.127-20.023z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowLeft);
export default ForwardRef;
