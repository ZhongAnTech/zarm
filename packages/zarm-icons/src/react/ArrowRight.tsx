import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowRight = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M711.156 508.035a41.472 41.472 0 01-11.126 20.023L366.12 861.97c-16.272 16.272-42.654 16.272-58.926 0s-16.272-42.654 0-58.926L611.767 498.47 307.241 193.944c-16.272-16.272-16.272-42.654 0-58.926 16.271-16.272 42.653-16.272 58.925 0L700.078 468.93c10.606 10.606 14.299 25.508 11.078 39.105z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowRight);
export default ForwardRef;
