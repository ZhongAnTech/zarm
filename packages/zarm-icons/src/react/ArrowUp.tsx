import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowUp = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M513.176 285.05a41.472 41.472 0 0120.024 11.127L867.11 630.089c16.272 16.271 16.272 42.653 0 58.925s-42.654 16.272-58.925 0L503.612 384.44 199.085 688.967c-16.272 16.271-42.654 16.271-58.925 0-16.272-16.272-16.272-42.654 0-58.926L474.07 296.129c10.606-10.606 25.508-14.299 39.105-11.078z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowUp);
export default ForwardRef;
