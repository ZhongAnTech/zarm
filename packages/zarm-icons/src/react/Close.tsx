import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgClose = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M560.149 501.223l255.344 255.344c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L501.223 560.149 245.88 815.493c-16.272 16.272-42.654 16.272-58.925 0-16.272-16.272-16.272-42.654 0-58.926l255.344-255.344L186.954 245.88c-16.272-16.272-16.272-42.654 0-58.925 16.271-16.272 42.653-16.272 58.925 0l255.344 255.344 255.344-255.344c16.272-16.272 42.654-16.272 58.926 0 16.272 16.271 16.272 42.653 0 58.925L560.149 501.223z"
          fillRule="nonzero"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgClose);
export default ForwardRef;
