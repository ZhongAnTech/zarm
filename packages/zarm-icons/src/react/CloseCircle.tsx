import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgCloseCircle = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <g fillRule="evenodd">
          <g>
            <path d="M500 460.716l137.493-137.493c10.848-10.848 28.436-10.848 39.284 0l.101.102c10.735 10.848 10.645 28.344-.203 39.08l-138.207 136.78 138.309 138.308c10.848 10.848 10.848 28.436 0 39.284-10.848 10.848-28.436 10.848-39.284 0L498.98 538.264 359.85 675.956c-10.41 10.303-27.188 10.26-37.545-.097-10.34-10.341-10.34-27.107 0-37.448L460.716 500 323.223 362.507c-10.848-10.848-10.848-28.436 0-39.284 10.848-10.848 28.436-10.848 39.284 0L500 460.716z" />
            <path
              d="M190.64 809.36c170.855 170.854 447.865 170.854 618.72 0 170.854-170.855 170.854-447.865 0-618.72-170.855-170.854-447.865-170.854-618.72 0-170.854 170.855-170.854 447.865 0 618.72zm-44.193 44.193c-195.263-195.262-195.263-511.844 0-707.106 195.262-195.263 511.844-195.263 707.106 0 195.263 195.262 195.263 511.844 0 707.106-195.262 195.263-511.844 195.263-707.106 0z"
              fillRule="nonzero"
            />
          </g>
        </g>
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgCloseCircle);
export default ForwardRef;
