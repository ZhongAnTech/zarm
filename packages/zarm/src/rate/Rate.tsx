import * as React from 'react';
import { createBEM } from '@zarm-design/bem';
import { StarFill } from '@zarm-design/icons';
import { useDrag } from '@use-gesture/react';
import { ConfigContext } from '../n-config-provider';
import { useControllableValue } from '../utils/hooks';
import { getBoundingClientRect } from '../utils/dom';
import type { BaseRateProps } from './interface';
import type { HTMLProps } from '../utils/utilityTypes';

export interface RateCssVars {
  '--size'?: number | string;
  '--color'?: React.CSSProperties['color'];
  '--active-color'?: React.CSSProperties['color'];
  '--gap'?: React.CSSProperties['marginRight'];
}

export type RateProps = BaseRateProps & HTMLProps;

const Rate = React.forwardRef<HTMLDivElement, RateProps>((props, ref) => {
  const { className, style, count, character, allowHalf, allowClear, readonly } = props;

  const [value, setValue] = useControllableValue<number>(props);
  const itemRefs = React.useRef<(HTMLDivElement | null)[]>([]);

  const { prefixCls } = React.useContext(ConfigContext);
  const bem = createBEM('rate', { prefixCls });

  const items = Array(count).fill(null);

  const getScoreRanges = () => {
    const ranges: { score: number; left: number }[] = [];
    itemRefs.current.forEach((item, index) => {
      if (!item) return;

      const rect = getBoundingClientRect(item);
      if (allowHalf) {
        ranges.push(
          { score: index + 0.5, left: rect.left },
          { score: index + 1, left: rect.left + rect.width / 2 },
        );
      } else {
        ranges.push({ score: index + 1, left: rect.left });
      }
    });

    return ranges;
  };

  const getScoreByPosition = (x: number) => {
    const ranges = getScoreRanges();
    for (let i = ranges.length - 1; i > 0; i--) {
      if (x > ranges[i].left) return ranges[i].score;
    }

    return allowHalf ? 0.5 : 1;
  };

  const render = (_, index: number) => {
    const score = index + 1;
    const checked = true;

    const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
      if (readonly) return;

      const next = getScoreByPosition(event.clientX);

      if (allowClear && value === next) {
        setValue(0);
        return;
      }

      setValue(allowHalf ? next : score);
    };

    const itemCls = bem('item', [
      {
        active: value >= index + 1,
        half: allowHalf && value === index + 0.5,
      },
    ]);

    return (
      <div
        key={index}
        ref={(element) => {
          itemRefs.current[index] = element;
        }}
        role="radio"
        tabIndex={0}
        className={itemCls}
        onClick={handleClick}
        aria-setsize={count}
        aria-posinset={score}
        aria-checked={checked}
      >
        <div className={bem('character__half')}>{character}</div>
        <div className={bem('character__full')}>{character}</div>
      </div>
    );
  };

  const cls = bem([{ readonly }, className]);

  const bind = useDrag(
    ({ values: [x] }) => {
      if (readonly) return;
      setValue(getScoreByPosition(x));
    },
    { pointer: { touch: true }, axis: 'x' },
  );
  return (
    <div ref={ref} role="radiogroup" tabIndex={0} className={cls} style={style} {...bind()}>
      {items.map(render)}
    </div>
  );
});

Rate.displayName = 'Rate';

Rate.defaultProps = {
  count: 5,
  defaultValue: 0,
  allowHalf: false,
  allowClear: false,
  character: <StarFill />,
};

export default Rate;
