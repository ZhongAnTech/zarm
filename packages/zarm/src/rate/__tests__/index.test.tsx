import * as React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Rate from '../index';

describe('Rate', () => {
  it('renders correctly', () => {
    const wrapper = render(<Rate />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should emit change and value event when rate icon is clicked', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Rate onChange={onChange} />);
    const item = wrapper.find('.za-rate__item');
    item.at(3).simulate('click');
    expect(onChange).toBeCalledWith(4);
  });
});
