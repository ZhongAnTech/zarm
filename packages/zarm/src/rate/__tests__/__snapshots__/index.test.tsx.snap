// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate renders correctly 1`] = `
<div
  class="za-rate"
  role="radiogroup"
  tabindex="0"
>
  <div
    aria-checked="true"
    aria-posinset="1"
    aria-setsize="5"
    class="za-rate__item"
    role="radio"
    tabindex="0"
  >
    <div
      class="za-rate__character"
    >
      <i
        class="za-icon za-icon--default za-icon--md"
        theme="default"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1000 1000"
          width="1em"
        >
          <path
            d="M500.122 852.313L222.594 974.61c-15.569 6.86-33.752-.199-40.612-15.767a30.806 30.806 0 01-2.46-15.526l30.551-301.737L8.001 415.427C-3.335 402.74-2.24 383.266 10.447 371.93a30.806 30.806 0 0114.005-7.136l296.41-64.187L473.502 38.54c8.563-14.702 27.423-19.678 42.125-11.115a30.806 30.806 0 0111.115 11.115l152.64 262.067 296.41 64.187c16.628 3.6 27.188 20 23.588 36.627a30.806 30.806 0 01-7.137 14.006L790.171 641.58l30.55 301.737c1.714 16.927-10.618 32.038-27.545 33.752-5.3.537-10.65-.31-15.526-2.459L500.122 852.313z"
            fill-rule="evenodd"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    aria-checked="true"
    aria-posinset="2"
    aria-setsize="5"
    class="za-rate__item"
    role="radio"
    tabindex="0"
  >
    <div
      class="za-rate__character"
    >
      <i
        class="za-icon za-icon--default za-icon--md"
        theme="default"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1000 1000"
          width="1em"
        >
          <path
            d="M500.122 852.313L222.594 974.61c-15.569 6.86-33.752-.199-40.612-15.767a30.806 30.806 0 01-2.46-15.526l30.551-301.737L8.001 415.427C-3.335 402.74-2.24 383.266 10.447 371.93a30.806 30.806 0 0114.005-7.136l296.41-64.187L473.502 38.54c8.563-14.702 27.423-19.678 42.125-11.115a30.806 30.806 0 0111.115 11.115l152.64 262.067 296.41 64.187c16.628 3.6 27.188 20 23.588 36.627a30.806 30.806 0 01-7.137 14.006L790.171 641.58l30.55 301.737c1.714 16.927-10.618 32.038-27.545 33.752-5.3.537-10.65-.31-15.526-2.459L500.122 852.313z"
            fill-rule="evenodd"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    aria-checked="true"
    aria-posinset="3"
    aria-setsize="5"
    class="za-rate__item"
    role="radio"
    tabindex="0"
  >
    <div
      class="za-rate__character"
    >
      <i
        class="za-icon za-icon--default za-icon--md"
        theme="default"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1000 1000"
          width="1em"
        >
          <path
            d="M500.122 852.313L222.594 974.61c-15.569 6.86-33.752-.199-40.612-15.767a30.806 30.806 0 01-2.46-15.526l30.551-301.737L8.001 415.427C-3.335 402.74-2.24 383.266 10.447 371.93a30.806 30.806 0 0114.005-7.136l296.41-64.187L473.502 38.54c8.563-14.702 27.423-19.678 42.125-11.115a30.806 30.806 0 0111.115 11.115l152.64 262.067 296.41 64.187c16.628 3.6 27.188 20 23.588 36.627a30.806 30.806 0 01-7.137 14.006L790.171 641.58l30.55 301.737c1.714 16.927-10.618 32.038-27.545 33.752-5.3.537-10.65-.31-15.526-2.459L500.122 852.313z"
            fill-rule="evenodd"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    aria-checked="true"
    aria-posinset="4"
    aria-setsize="5"
    class="za-rate__item"
    role="radio"
    tabindex="0"
  >
    <div
      class="za-rate__character"
    >
      <i
        class="za-icon za-icon--default za-icon--md"
        theme="default"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1000 1000"
          width="1em"
        >
          <path
            d="M500.122 852.313L222.594 974.61c-15.569 6.86-33.752-.199-40.612-15.767a30.806 30.806 0 01-2.46-15.526l30.551-301.737L8.001 415.427C-3.335 402.74-2.24 383.266 10.447 371.93a30.806 30.806 0 0114.005-7.136l296.41-64.187L473.502 38.54c8.563-14.702 27.423-19.678 42.125-11.115a30.806 30.806 0 0111.115 11.115l152.64 262.067 296.41 64.187c16.628 3.6 27.188 20 23.588 36.627a30.806 30.806 0 01-7.137 14.006L790.171 641.58l30.55 301.737c1.714 16.927-10.618 32.038-27.545 33.752-5.3.537-10.65-.31-15.526-2.459L500.122 852.313z"
            fill-rule="evenodd"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    aria-checked="true"
    aria-posinset="5"
    aria-setsize="5"
    class="za-rate__item"
    role="radio"
    tabindex="0"
  >
    <div
      class="za-rate__character"
    >
      <i
        class="za-icon za-icon--default za-icon--md"
        theme="default"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1000 1000"
          width="1em"
        >
          <path
            d="M500.122 852.313L222.594 974.61c-15.569 6.86-33.752-.199-40.612-15.767a30.806 30.806 0 01-2.46-15.526l30.551-301.737L8.001 415.427C-3.335 402.74-2.24 383.266 10.447 371.93a30.806 30.806 0 0114.005-7.136l296.41-64.187L473.502 38.54c8.563-14.702 27.423-19.678 42.125-11.115a30.806 30.806 0 0111.115 11.115l152.64 262.067 296.41 64.187c16.628 3.6 27.188 20 23.588 36.627a30.806 30.806 0 01-7.137 14.006L790.171 641.58l30.55 301.737c1.714 16.927-10.618 32.038-27.545 33.752-5.3.537-10.65-.31-15.526-2.459L500.122 852.313z"
            fill-rule="evenodd"
          />
        </svg>
      </i>
    </div>
  </div>
</div>
`;
